### PR TITLE
BLS signature verification

### PR DESCRIPTION
**Pull request by janezpodhostnik**
closes: https://github.com/dapperlabs/flow-go/issues/5544

### CODE REVIEW

**Comment by tarakby** [comment]
Another question: what is the difference between these 2 verification use cases? I mean I see they are interfaces differences, but isn't one enough?
- https://github.com/onflow/flow-go/blob/janez/bls-sig-verification/fvm/crypto/crypto.go#L18
- https://github.com/onflow/flow-go/blob/janez/bls-sig-verification/fvm/crypto/crypto.go#L164

**Review by ramtinms — APPROVED** [review]

File: fvm/crypto/crypto.go [hunk 1]
```diff
@@ -172,8 +174,14 @@ func VerifySignatureFromRuntime(
...
 	}
-	if sigAlgo == crypto.BLSBLS12381 {
-		return false, errors.NewValueErrorf(signatureAlgorithm.Name(), "signature algorithm type %s not supported", crypto.BLSBLS12381.String())
+
+	tag := parseRuntimeDomainTag(rawTag)
+	if tag == nil {
+		return false, errors.NewValueErrorf(rawTag, "invalid domain tag")
+	}
+
+	if len(tag) > 0 && sigAlgo != crypto.BLSBLS12381 && string(tag) != string(flow.UserDomainTag[:]) {
```
[the remark this answers is not in the document]
Comment (janezpodhostnik) [reply fvm/crypto/crypto.go:183]: You are right, there is a condition above that makes sure tag is not empty.

**Comment by janezpodhostnik** [comment]
@tarakby https://github.com/onflow/flow-go/blob/janez/bls-sig-verification/fvm/crypto/crypto.go#L164 calls https://github.com/onflow/flow-go/blob/janez/bls-sig-verification/fvm/crypto/crypto.go#L18. The `Verify` could be in-lined and the `SignatureVerifier` interface removed, but i think this way it might be easier to test.

File: fvm/crypto/crypto.go [hunk 2]
```diff
@@ -172,8 +174,14 @@ func VerifySignatureFromRuntime(
 	if sigAlgo == crypto.UnknownSigningAlgorithm {
 		return false, errors.NewValueErrorf(signatureAlgorithm.Name(), "signature algorithm type not found")
 	}
-	if sigAlgo == crypto.BLSBLS12381 {
-		return false, errors.NewValueErrorf(signatureAlgorithm.Name(), "signature algorithm type %s not supported", crypto.BLSBLS12381.String())
+
+	tag := parseRuntimeDomainTag(rawTag)
```
[the remark this answers is not in the document]
Comment (janezpodhostnik) [reply fvm/crypto/crypto.go]: The second option looks better to me. I will open an issue so we don't forget to take this compatibility out.

**Pull request merged**